README.md
Rate Limit Memcached
====================

Memcached client for the [express-rate-limit](https://github.com/nfriedly/express-rate-limit) middleware.

<a href="https://www.npmjs.com/package/rate-limit-memcached" title="npm"><img src="http://img.shields.io/npm/v/rate-limit-memcached.svg?style=for-the-badge"></a>
<a href="https://travis-ci.org/linyows/rate-limit-memcached" title="travis"><img src="https://img.shields.io/travis/linyows/rate-limit-memcached.svg?style=for-the-badge"></a>
<a href="https://coveralls.io/github/linyows/rate-limit-memcached" title="coveralls"><img src="https://img.shields.io/coveralls/linyows/rate-limit-memcached.svg?style=for-the-badge"></a>
<a href="https://github.com/linyows/rate-limit-memcached/blob/master/MIT-LICENSE" title="MIT License"><img src="https://img.shields.io/badge/license-MIT-blue.svg?style=for-the-badge"></a>

Installation
------------

```sh
$ npm install --save rate-limit-memcached
```

Usage
-----

### Node.js

```js
var RateLimit = require('express-rate-limit');
var Memcached = require('memcached');
var MemcachedStore = require('rate-limit-memcached');

var client = new Memcached(['memcached.example.com:11211']);
var expiration = 20 * 60;
var limiter = new RateLimit({
  store: new MemcachedStore({ expiration: expiration, client: client, prefix: 'remoteip:' }),
  windowMs: expiration * 1000,
  max: 500,
  delayMs: 0
});

app.use(limiter);
```

### TypeScript

```ts
import * as RateLimit from 'express-rate-limit'
import * as Memcached from 'memcached'
import MemcachedStore from 'rate-limit-memcached'

const client = new Memcached(['memcached.example.com:11211'])
const expiration = 20 * 60
const limiter = new RateLimit({
  store: new MemcachedStore({ expiration, client, prefix: 'remoteip:' }),
  windowMs: expiration * 1000,
  max: 500,
  delayMs: 0
})

app.use(limiter)
```

Contribution
------------

1. Fork (https://github.com/linyows/rate-limit-memcached/fork)
1. Create a feature branch
1. Commit your changes
1. Rebase your local changes against the master branch
1. Install dependencies with the `npm ci` command
1. Run test suite with the `npm test` command and confirm that it passes
1. Create a new Pull Request

Author
------

[linyows](https://github.com/linyows)


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.0.1](https://github.com/express-rate-limit/rate-limit-redis/releases/tag/v1.0.1)

### Added

- Enabled provenance statement generation, see
  https://github.com/express-rate-limit/express-rate-limit#406.

## [1.0.0](https://github.com/express-rate-limit/rate-limit-memcached/releases/tag/v1.0.0)

### Breaking

- Rewrite the store to implement the new `Store` interface, introduced in
  `express-rate-limit` v6.0.0.
- Require the `del`, `get`, `set`, `add`, `incr` and `decr` functions on any
  `memcached` client passed to the store.
- `MemcachedStore` is now a named export, not a default export, to make it play
  nice with the dual cjs-esm package.

### Added

- Add the `locations` and `config` options.
- Store expiry time for a client in a new key, named `${prefix}expiry:${key}`.


test/types.ts
// /test/types.ts
// Setup the types for `memcached-mock`

declare module 'memcached-mock' {
	export { default } from 'memcached'
}


test/store-test.ts
// /test/store-test.ts
// The tests for the increment, decrement and resetKey functions.

import Memcached from 'memcached-mock'
import { it, expect, jest } from '@jest/globals'
import { MemcachedStore } from '../source/index.js'
import './types.js' // eslint-disable-line import/no-unassigned-import

// Return the same spied-on instance of the store for the tests.
const getStore = async (): Promise<MemcachedStore> => {
	// Create a new client instance, and clear its cache. Somehow the cache from
	// the previous test carries over unless we call flush here.
	const client = new Memcached('localhost:11211')
	await new Promise<void>((resolve, reject) => {
		client.flush((error) => {
			if (error) {
				reject(error)
				return
			}

			resolve()
		})
	})

	// Spy on all the functions so we can make sure they are called.
	jest.spyOn(client, 'get')
	jest.spyOn(client, 'set')
	jest.spyOn(client, 'add')
	jest.spyOn(client, 'incr')
	jest.spyOn(client, 'decr')
	jest.spyOn(client, 'del')

	// Create an new store and initialise it.
	const store = new MemcachedStore({ client })
	// @ts-expect-error We only need to pass the `windowMs` option.
	store.init({ windowMs: 2 * 1000 })
	return store
}

it('should work when `increment` is called for new key', async () => {
	const store = await getStore()
	const data = await store.increment('1.2.3.4')

	expect(data.totalHits).toBe(1)
	expect(data.resetTime instanceof Date).toBe(true)

	expect(store.client.incr).toHaveBeenCalled()
	expect(store.client.add).toHaveBeenCalledTimes(2)
})

it('should work when `increment` is called for existing key', async () => {
	const store = await getStore()

	await store.increment('1.2.3.4')
	await store.increment('1.2.3.4')

	const data = await store.increment('1.2.3.4')

	expect(data.totalHits).toBe(3)
	expect(data.resetTime instanceof Date).toBe(true)
})

it('should count all hits when `increment` is called simultaneously', async () => {
	const store = await getStore()

	await Promise.all([
		store.increment('1.2.3.4'),
		store.increment('1.2.3.4'),
		store.increment('1.2.3.4'),
	])

	const data = await store.increment('1.2.3.4')

	expect(data.totalHits).toBe(4)
	expect(data.resetTime instanceof Date).toBe(true)
})

it('should still call `decr` when `decrement` is called for non-existent key', async () => {
	const store = await getStore()

	await store.decrement('1.2.3.4')

	expect(store.client.decr).toHaveBeenCalled()
})

it('should work when `decrement` is called for existing key', async () => {
	const store = await getStore()

	await store.increment('1.2.3.4')
	await store.decrement('1.2.3.4')

	expect(store.client.decr).toHaveBeenCalled()
})

it('should work when `resetKey` is called for existing key', async () => {
	const store = await getStore()

	await store.increment('1.2.3.4')
	await store.resetKey('1.2.3.4')

	expect(store.client.del).toHaveBeenCalledTimes(2) // Once for the key, once for the `key__expiry`.
})

it('should still call `del` when `resetKey` is called for non-existent key', async () => {
	const store = await getStore()

	await store.resetKey('1.2.3.4')

	expect(store.client.del).toHaveBeenCalledTimes(2)
})


source/store.ts
// /source/store.ts
// A `memcached` store for the `express-rate-limit` middleware.

import type {
	Store,
	Options as RateLimitConfiguration,
	IncrementResponse,
} from 'express-rate-limit'
import Memcached from 'memcached'
import type { Options, MemcachedClient } from './types'

/**
 * The hit count and start window timestamp of a client.
 */
type ClientRecord = { hits: number; time: number } // [in milliseconds]

/**
 * A `Store` for the `express-rate-limit` package that stores hit counts in
 * Memcached.
 */
class MemcachedStore implements Store {
	/**
	 * The number of milliseconds to remember that user's requests.
	 */
	expiration!: number

	/**
	 * The text to prepend to the key.
	 */
	prefix!: string

	/**
	 * The `memcached` client to use.
	 */
	client!: MemcachedClient

	/**
	 * @constructor for `MemcachedStore`.
	 *
	 * @param options {Options} - The options used to configure the store's behaviour.
	 */
	constructor(options?: Partial<Options>) {
		this.prefix = options?.prefix ?? 'rl:'

		if (options?.client) {
			if (
				typeof options.client.get === 'function' &&
				typeof options.client.add === 'function' &&
				typeof options.client.replace === 'function' &&
				typeof options.client.del === 'function'
			)
				this.client = options.client
			else throw new Error('An invalid memcached client was passed to store.')
		} else
			this.client = new Memcached(
				options?.locations ?? ['localhost:11211'],
				options?.config ?? {},
			)
	}

	/**
	 * Method that actually initializes the store.
	 *
	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
	 *
	 * @impl
	 */
	init(options: RateLimitConfiguration) {
		this.expiration = options.windowMs / 1000 // [in seconds]
	}

	/**
	 * Method to prefix the keys with the given text.
	 *
	 * @param key {string} - The key.
	 *
	 * @returns {string} - The text + the key.
	 */
	prefixKey(key: string): string {
		return `${this.prefix}${key}`
	}

	/**
	 * Method to increment a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client.
	 *
	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
	 */
	async increment(key: string): Promise<IncrementResponse> {
		const response = (await this.updateKey(
			this.prefixKey(key),
			+1,
		)) as IncrementResponse // We can cast this for sure because we're passing delta as +1
		return response
	}

	/**
	 * Method to decrement a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 */
	async decrement(key: string): Promise<void> {
		await this.updateKey(this.prefixKey(key), -1)
	}

	/**
	 * Method to reset a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client.
	 */
	async resetKey(key: string): Promise<void> {
		const prefixedKey = this.prefixKey(key)

		// Delete the record for that key.
		await new Promise<void>((resolve, reject) => {
			this.client.del(prefixedKey, (error) => {
				if (error) {
					reject(error)
					return
				}

				resolve()
			})
		})
	}

	private async updateKey(
		key: string,
		delta: number,
	): Promise<IncrementResponse | void> {
		// Find the existing record for that key.
		let record = await new Promise<ClientRecord>((resolve, reject) => {
			this.client.get(key, (error, data) => {
				if (error) {
					reject(error)
					return
				}

				resolve(data as ClientRecord)
			})
		})

		if (record === undefined && delta < 0) {
			// If the record does not exist, and we are supposed to decrement the key,
			// we don't need to do anything, so we return.
			return
		}

		if (record === undefined) {
			// If a record does not exist, and must be incremented, then add a record
			// for that key.
			record = await new Promise<ClientRecord>((resolve, reject) => {
				const payload: ClientRecord = { hits: 1, time: Date.now() }

				this.client.add(key, payload, this.expiration, (error) => {
					if (error) {
						reject(error)
						return
					}

					resolve(payload)
				})
			})
		} else {
			// If it does exist, simply change the hit count by `delta`.
			record = await new Promise<ClientRecord>((resolve, reject) => {
				const payload: ClientRecord = { ...record, hits: record.hits + delta }
				const expiration = Math.floor(
					// = window - (time elapsed since start of window) [in seconds]
					this.expiration - (Date.now() - payload.time) / 1000,
				)

				this.client.replace(key, payload, expiration, (error) => {
					if (error) {
						reject(error)
						return
					}

					resolve(payload)
				})
			})
		}

		// Return the total number of hits and the time left for the window to reset.
		return {
			totalHits: record.hits,
			resetTime: new Date(record.time + this.expiration * 1000),
		}
	}
}

// Export it to the world!
export default MemcachedStore


source/types.ts
// /source/types.ts
// The type definitions for this package.

import type Memcached from 'memcached'

/**
 * A memcached client.
 */
export type MemcachedClient = {
	get: (key: string, cb: (error: any, data: any) => void) => void
	set: (key: string, value: any, time: number, cb: (error: any) => void) => void
	add: (key: string, value: any, time: number, cb: (error: any) => void) => void
	del: (key: string, cb: (error: any) => void) => void
	incr: (key: string, amount: number, cb: (error: any) => void) => void
	decr: (key: string, amount: number, cb: (error: any) => void) => void
}

/**
 * The configuration options for the store.
 */
export type Options = {
	/**
	 * The text to prepend to the key.
	 */
	prefix: string

	/**
	 * The `memcached` client to use.
	 */
	client: MemcachedClient

	/**
	 * A list of memcached server URLs to store the keys in, passed to the default
	 * memcached client.
	 *
	 * Note that the default client is only used if another client is not passed
	 * to the store.
	 */
	locations: string[]

	/**
	 * The configuration to pass to the default client, along with the `locations`.
	 */
	config: Memcached.options
}


source/memcached-store.ts
// /source/memcached-store.ts
// A `memcached` store for the `express-rate-limit` middleware.

import { promisify } from 'node:util'
import type {
	Store,
	Options as RateLimitConfiguration,
	IncrementResponse,
} from 'express-rate-limit'
import Memcached from 'memcached'
import type { Options, MemcachedClient } from './types'

// A list of methods that should be present on a client object.
const methods: Array<keyof MemcachedClient> = [
	'del',
	'get',
	'set',
	'add',
	'incr',
	'decr',
]

/**
 * The promisifed version of the `MemcachedClient`.
 */
type PromisifiedMemcachedClient = {
	get: <T>(key: string) => Promise<T | undefined>
	set: (key: string, value: any, time: number) => Promise<void>
	add: (key: string, value: any, time: number) => Promise<void>
	del: (key: string) => Promise<boolean | undefined>
	incr: (key: string, amount: number) => Promise<boolean | number>
	decr: (key: string, amount: number) => Promise<boolean | number>
}

/**
 * A `Store` for the `express-rate-limit` package that stores hit counts in
 * Memcached.
 */
class MemcachedStore implements Store {
	/**
	 * The number of seconds to remember a client's requests.
	 */
	expiration!: number

	/**
	 * The text to prepend to the key.
	 */
	prefix!: string

	/**
	 * The `memcached` client to use.
	 */
	client!: MemcachedClient

	/**
	 * The promisifed functions from the `client` object.
	 */
	fns!: PromisifiedMemcachedClient

	/**
	 * @constructor for `MemcachedStore`.
	 *
	 * @param options {Options} - The options used to configure the store's behaviour.
	 */
	constructor(options?: Partial<Options>) {
		this.prefix = options?.prefix ?? 'rl:'

		if (options?.client) {
			for (const func of methods) {
				if (typeof options.client[func] !== 'function')
					throw new Error('An invalid memcached client was passed to store.')
			}

			this.client = options.client
		} else {
			this.client = new Memcached(
				options?.locations ?? ['localhost:11211'],
				options?.config ?? {},
			)
		}

		// Promisify the functions.
		// @ts-expect-error This line simply initialises the object, calm down lol.
		this.fns = {}
		for (const func of methods) {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			this.fns[func] = promisify(this.client[func]).bind(this.client)
		}
	}

	/**
	 * Method that actually initializes the store.
	 *
	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
	 *
	 * @impl
	 */
	init(options: RateLimitConfiguration) {
		this.expiration = options.windowMs / 1000 // [in seconds]
	}

	/**
	 * Method to prefix the keys with the given text.
	 *
	 * @param key {string} - The key.
	 *
	 * @returns {string} - The text + the key.
	 */
	prefixKey(key: string): string {
		return `${this.prefix}${key}`
	}

	/**
	 * Method that returns the name of the key used to store the reset timestamp
	 * for the given key.
	 *
	 * @param key {string} - The key.
	 *
	 * @returns {string} - The expiry key's name.
	 */
	expiryKey(key: string): string {
		return `${this.prefix}expiry:${key}`
	}

	/**
	 * Method to increment a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client.
	 *
	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
	 */
	async increment(key: string): Promise<IncrementResponse> {
		const prefixedKey = this.prefixKey(key)

		// Try incrementing the given key. If the key exists, it will increment it
		// and return the updated hit count.
		let totalHits = await this.fns.incr(prefixedKey, 1)
		let expiresAt

		if (totalHits === false) {
			try {
				// The increment command failed since the key does not exist. In which case, set the
				// hit count for that key to 1, and make sure it expires after `window` seconds.
				await this.fns.add(prefixedKey, 1, this.expiration)

				// If it is added successfully, set `totalHits` to 1.
				totalHits = 1

				// Also store the expiration time in a separate key.
				expiresAt = Date.now() + this.expiration * 1000 // [seconds -> milliseconds]
				await this.fns.add(
					this.expiryKey(key), // The name of the key.
					expiresAt, // The value - the time at which the key expires.
					this.expiration, // The key should be deleted by memcached after `window` seconds.
				)
			} catch (caughtError: any) {
				const error = caughtError as Error

				// If the `add` operation fails because the key already exists, it was
				// created sometime in between, call `increment` again, and fetch its
				// expiry time.
				if (/not(\s)?stored/i.test(error.message)) {
					totalHits = await this.fns.incr(prefixedKey, 1)
					expiresAt = await this.fns.get<number>(this.expiryKey(key))
				} else {
					// Otherwise, throw the error.
					throw error
				}
			}
		} else {
			// If the key exists and has been incremented succesfully, retrieve its expiry.
			expiresAt = await this.fns.get<number>(this.expiryKey(key))
		}

		// Make sure `totalHits` is a number.
		if (typeof totalHits !== 'number')
			throw new Error(
				`Expected 'totalHits' to be a number, got ${totalHits} instead.`,
			)

		// Return the total number of hits, as well as the reset timestamp.
		return {
			totalHits,
			// If `expiresAt` is undefined, assume the key expired sometime in between
			// reading the hits and expiry keys from memcached.
			resetTime: expiresAt ? new Date(expiresAt) : new Date(),
		}
	}

	/**
	 * Method to decrement a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 */
	async decrement(key: string): Promise<void> {
		// Decrement the key, and do nothing if it doesn't exist.
		await this.fns.decr(this.prefixKey(key), 1)
	}

	/**
	 * Method to reset a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client.
	 */
	async resetKey(key: string): Promise<void> {
		// Delete the the key, as well as its expiration counterpart.
		await this.fns.del(this.prefixKey(key))
		await this.fns.del(this.expiryKey(key))
	}
}

// Export it to the world!
export default MemcachedStore
